### PR TITLE
HBJSON -> GIF / Transparent Images

### PR DESCRIPTION
**Pull request by devngc**
Hi @mostaphaRoudsari,

Making this PR happen finally. I am sure this PR will convince you that honeybee-vtk is in need of a refactor. Following are the things I think should do next;

1. This workflow has two parts. 1. Generate time step images. 2. Process those images to generate GIF and transparent images. Step 1 can take really long based on the grid density. Therefore, I would like to break the command or else if someone wants to tweak something in image processing, they have to re-run the step to generate time step images. This will be really bad.
2. I need to generally test the command more.
3. I need to implement parallel processing for the step-1

### CODE REVIEW

File: honeybee_vtk/assistant.py [hunk 1]
```diff
@@ -37,6 +37,16 @@ def __init__(self, background_color: Tuple[int, int, int], camera: Camera,
...
+        self._vtk_camera = None
+        self._active_camera = None
+
+    @property
+    def vtk_camera(self) -> vtk.vtkCamera:
+        return self._vtk_camera
+
+    @vtk_camera.setter
+    def vtk_camera(self, vtk_camera):
+        self._vtk_camera = vtk_camera
```
Comment (mostaphaRoudsari) [inline honeybee_vtk/assistant.py:49]: We probably need a check here to ensure it is the correct type.

File: honeybee_vtk/assistant.py [hunk 2]
```diff
@@ -196,8 +209,8 @@ def auto_text_height(self, image_width, image_height) -> None:
     def _export_image(
             self, folder: str, image_type: ImageTypes = ImageTypes.png, *,
             image_scale: int = 1, image_width: int = 0, image_height: int = 0,
-            color_range: vtk.vtkLookupTable = None, rgba: bool = False,
-            show: bool = False) -> str:
+            image_name: str = '',  color_range: vtk.vtkLookupTable = None,
```
Comment (mostaphaRoudsari) [inline honeybee_vtk/assistant.py:212]: Setting the default to '' is not a good idea. Set them to None.

File: honeybee_vtk/assistant.py [hunk 3]
```diff
@@ -37,6 +37,16 @@ def __init__(self, background_color: Tuple[int, int, int], camera: Camera,
         self._camera = camera
         self._legend_params = legend_parameters
         self._text_actor = text_actor
+        self._vtk_camera = None
+        self._active_camera = None
```
Comment (mostaphaRoudsari) [inline honeybee_vtk/assistant.py:41]: Why are you adding `_active_camera` here if you never using it in the class?

File: honeybee_vtk/cli/export.py [hunk 4]
```diff
@@ -107,11 +112,56 @@ def export():
...
+)
+@click.option(
+    '--flat/--gradient', default=True, show_default=True, help='Indicate'
+    ' whether to use flat transparency or gradient transparency in generating time step'
+    ' images.'
+)
+@click.option(
+    '--transparent-images', '-ti', is_flag=True, default=False, show_default=True,
+    help='Indicate whether to generate transparent images or not.'
+)
```
Comment (mostaphaRoudsari) [inline honeybee_vtk/cli/export.py:135]: To be able to use the flags with Pollination we need them to have two options.

File: honeybee_vtk/config.py [hunk 5]
```diff
@@ -229,3 +250,38 @@ class Config(BaseModel):
...
+    """A DateTime object."""
+    month: conint(ge=1, le=12) = Field(
+        description='Month of the year. Accepted values are between 1 and 12 inclusive.'
+    )
+    day: conint(ge=1, le=31) = Field(
+        description='Month of the year. Accepted values are between 1 and 31 inclusive.'
+    )
+    hour: conint(ge=0, le=23) = Field(
+        description='Hour of the day. Accepted values are between 0 and 23 inclusive.'
+    )
```
Comment (mostaphaRoudsari) [inline honeybee_vtk/config.py:265]: How do we handle timesteps smaller than one?

File: honeybee_vtk/config.py [hunk 6]
```diff
@@ -229,3 +250,38 @@ class Config(BaseModel):
         description='List of data-config objects that define the data to be loaded on'
         ' the model.'
     )
+
+
+class DateTimeConfig(BaseModel):
```
Comment (mostaphaRoudsari) [inline honeybee_vtk/config.py:255]: I'm not sure why do you need this object? Why don't you use Ladybug's DateTime type directly?

File: honeybee_vtk/legend_parameter.py [hunk 7]
```diff
@@ -571,6 +572,15 @@ def range(self) -> Tuple[float, float]:
                     f' value {self._min}.'
                 )
 
+    def _assign_colors(self, colors: Tuple[Color, Color]) -> None:
```
Comment (mostaphaRoudsari) [inline honeybee_vtk/legend_parameter.py:575]: From the code, it looks like you want a Tuple with variable-length but you are setting it here to a Tuple with two items.
https://docs.python.org/3/library/typing.html#typing.Tuple

**Review by mostaphaRoudsari — COMMENTED** [review]
Hi @devngc!

See my comments. I didn't review every line of the code as this is too many changes to be reviewed but left some random comments for some red flags that I saw.

File: honeybee_vtk/_helper.py [hunk 8]
```diff
@@ -53,3 +56,25 @@ def get_line_count(file_path: str) -> int:
...
+    """Decorator to measure time used by a function."""
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        """This is a wrapper function."""
+        start = time.perf_counter()
+        result = func(*args, **kwargs)
+        finish = time.perf_counter()
+        print(f'Finished in {report_time(finish-start)}')
+        return result
+    return wrapper
```
[the remark this answers is not in the document]
Comment (devngc) [reply honeybee_vtk/_helper.py:80]: I removed it.

File: honeybee_vtk/assistant.py (hunk 1 above)
Comment (devngc) [reply honeybee_vtk/assistant.py:49]: I did that. This is one of those things that we go back and forth on. I tend to write in what is called defensive programming style where you set up checks for such types. In many of the earlier comments, you have not liked it. I would agree that every situation is different and we cannot have a blanket rule though.

File: honeybee_vtk/assistant.py (hunk 2 above)
Comment (devngc) [reply honeybee_vtk/assistant.py:212]: Yes. this is my bad.

File: honeybee_vtk/assistant.py (hunk 3 above)
Comment (devngc) [reply honeybee_vtk/assistant.py:41]: It's a remnant of an old design I was working on. I removed it.

File: honeybee_vtk/cli/export.py (hunk 4 above)
Comment (devngc) [reply honeybee_vtk/cli/export.py:135]: Noted.  Turned it into a flag with two options.

File: honeybee_vtk/config.py (hunk 5 above)
Comment (devngc) [reply honeybee_vtk/config.py:265]: So the way thresholding will work is, all the timesteps between two DateTimes will be caught. Say for example, the time step is one hour and DateTime is (6,21,8) to (6.21,9) then 1 timestep will be caught. If the time steps if 0.5, two timesteps will be caught.

File: honeybee_vtk/config.py (hunk 6 above)
Comment (devngc) [reply honeybee_vtk/config.py:255]: I don't understand  your comment here. How can you make a Ladybug tools DateTime object in json file?

**Review by mostaphaRoudsari — APPROVED** [review]
Thanks! Let's merge it in.

**Pull request merged**